GFF, CSV validator and Documentation

deepakkumar96 opened the pull request:
- Add Gff3 Validator
- Add CSV validator
    - add dependency to 'univocity-parsers' for parsing csv/tsv files
- Add Documentation and update release_note
- Add restriction on the maximum number of warnings that can be generated by the validator

Issues resolved:
- Always return validation-result even if an exception is thrown, Fixes #25
- Guess file type from filename if not provided, Fixes #24
- Verbose error messages in-case file not found or invalid type. Fixes #19
- Validation Rules are add in the GitHub wiki pages and support for validation rules in the command line validator, Fixes #18
- Added validationResult.getErrorMessage() to return single error message, Fixes #13
-  Had refactored FastaValidator::extractSequenceIdFromHeader(), Fixes #6

deepakkumar96 committed:
Refactor Fasta validator
- move Fasta Header validation a separate method

deepakkumar96 committed:
Merge branch 'validator-api-feat' into feat-gff3-validator

deepakkumar96 committed:
Add a Gff3 Feature Parser

deepakkumar96 committed:
Creat a simple Gff3 Validator.

deepakkumar96 committed:
Add Gff3 validation rules

deepakkumar96 committed:
Add parent and unique id validation to Gff3 validator

deepakkumar96 committed:
Add Gff3 score and coordinate validation rules:
- Change score type in Gff3Feature from long to String

deepakkumar96 committed:
Add Gff3 phase and strands validation rules:

deepakkumar96 committed:
Add single feature attribute validation

deepakkumar96 committed:
Add Gff3 Validator in biovalidator:
- Add getErrorMessage() method to return single error message
  in ValidationResult. This closes #13.
- Add GFF3 type in ValidatorType
- Add a helper method in ValidatorHelper for Gff3 validation
- Fix Gff errors and add few more feature attribute validation
- Add few basic Gff3 tests

deepakkumar96 committed:
Add master build status image in Readme.MD

deepakkumar96 committed:
Fix CDS feature check type!

deepakkumar96 committed:
Fix seqId validation with respect to ##sequence-region directive

deepakkumar96 committed:
Add #sequence-region pragma check rule

deepakkumar96 committed:
Fix start/end validation, start must be greater than 0.

deepakkumar96 committed:
Add Test for Gff3 comment directives and Fix passing invalid version

deepakkumar96 committed:
Add Test for Gff3 SeqId

deepakkumar96 committed:
Fix seqId validation to allow escaped '>' and spaces.
- add tests for seqId validation
- add tests for start and end coordinate validation

deepakkumar96 committed:
Add SequenceOntology check and Tests

deepakkumar96 committed:
Update biovalidator command-line utility and readme.md

deepakkumar96 committed:
Add score, strand and phase test

deepakkumar96 committed:
Update biovalidator command-line utility and readme.md

deepakkumar96 committed:
Add tests and error message validation
- also so.obo with so-simple.obo

deepakkumar96 committed:
Add missing processed_so_terms.obo file

deepakkumar96 committed:
Add tests for parentOf relation and attribute validation

deepakkumar96 committed:
Refactor GFF3 validator:
- now using pre-compiled regex patterns instead of compiling each pattern on runtime

- Instead of writing all validation rules within the same GFF3 class,
  now one or more rules are separated into a single class(called RuleValidators)
  and GFF3 validator can call validateAndAddError() method on those RuleValidator class

- Add another version of addError(str: String) and addWarning in ValidationResult

- Add a getter for validationResultStrategy in ValidationResult

deepakkumar96 committed:
Refactor GFF3 validator code

deepakkumar96 committed:
Merge remote-tracking branch 'upstream/master' into feat-gff3-validator

# Conflicts:
#	build.gradle

deepakkumar96 committed:
update build.gradle from upstream/master and remove invalid characters
from javadoc

deepakkumar96 committed:
Add csv/tsv validator:
- add dependency to 'univocity-parsers' for parsing csv/tsv files
- using 'edit distance' as the technique for measuring inconsistency
  of data in csv/tsv files

deepakkumar96 committed:
add csv validator type in CommandLineRunner

deepakkumar96 committed:
Add docs and RELEASE_NOTES

deepakkumar96 committed:
Fix start and end coordinate bug when shoudStopOnError is enabled:
- Add tests for source(column 2) and feature-type(column 3)
- Add tests for un-parsable start and end coordinates(column 4 and 5)
- Fix typo in 'GFF3FeatureTypeRuleValidator' class name
- Fix type in 'GFF3StartAndEndCoordinateRuleValidator' class name

deepakkumar96 committed:
Merge branch 'feat-csv-validator' into gff-and-csv

deepakkumar96 committed:
Add csv header detector

deepakkumar96 committed:
improve csv header detector and add csv column validation

deepakkumar96 committed:
remove extra characters in comment

deepakkumar96 committed:
Add improved version of csv consistency detector
- add isNotValid() method in ValidationResult

deepakkumar96 committed:
Add random pattern support in csv column consistency check

deepakkumar96 committed:
Fix check style wrong indentation error

deepakkumar96 committed:
add pattern creation of csv column

deepakkumar96 committed:
add few comments to explain code behaviour

deepakkumar96 committed:
Add repetitive pattern detection and improve CSV validator performance

deepakkumar96 committed:
fix wrong pattern creation and Pattern test
- refactor CsvColumnPattern
- Add dependency to apache-common-text

deepakkumar96 committed:
Refactor CsvSchemaValidator and improve csv validator performance

deepakkumar96 committed:
Solved bio-validator issues:
- Remove validation exception
- guess file types

deepakkumar96 committed:
Merge remote-tracking branch 'origin/docs-and-release' into gff-and-csv

deepakkumar96 committed:
Add javadoc for csv validator:
- Add --docs option support in command-line-runner
- Add BioValidatorUtils class for common methods

deepakkumar96 committed:
Add max pattern length constraints on CsvColumnPattern:
- Add more tests for csv validator
- refactor csv validator

deepakkumar96 committed:
Fix CsvValidator validator stream unclosed error and common issues:
- Fix close() method of CsvParser to auto-close input-stream
- Add verbose error message when parser fail to parse the file
- Fix typo('fatsa') in CommandLineRunner
- Change invalid validator type error message

deepakkumar96 committed:
Update RELEASE_NOTES to include GFF and CSV changes and bug fixes.

deepakkumar96 committed:
Refactor and documented csv validator

deepakkumar96 committed:
Add strict and permissive validation in csv validator

deepakkumar96 committed:
Add maximum number of warnings that can be generated by validator

deepakkumar96 committed:
Add CSV validator docs:
- change CSV strict validator similarity ratio to 0.9 from 0.85

deepakkumar96 committed:
Change biovalidator version from '0.1.0' to '0.1.2'

julie-sullivan commented:
Should the other pull requests be closed? as this is a replacement?

julie-sullivan commented:
(PS and all these changes look really great!) :)

deepakkumar96 commented:
Yes, All other PR should be closed, I didn't close just because you had
raised some issues and you may want to refer them,
"GFF, CSV validator and Documentation #26" is the one with all changes.

On Thu, Aug 8, 2019 at 8:53 AM Julie Sullivan <notifications@github.com>
wrote:

> (PS and all these changes look really great!) :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/biovalidator/pull/26?email_source=notifications&email_token=AED7EHCALF6JRFQMSZMAPX3QDPNJJA5CNFSM4IJGLJGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD326GVI#issuecomment-519430997>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AED7EHFOF4P6H4ESE376JCDQDPNJJANCNFSM4IJGLJGA>
> .
>

deepakkumar96 committed:
Add docs on custom gradle tasks and docs url in README.MD

julie-sullivan commented on README.md:
```diff
@@ -62,4 +62,35 @@ Validator validator = new FastaValidator(isr, SequenceType.DNA);
+
+### Using biovalidator as a command line utility
+Step 1: Get biovalidator Fat-Jar(either build from source or dirctly download from [bintray](https://bintray.com/intermineorg/biovalidator/biovalidator#files/org%2Fintermine%2Fbiovalidator%2F0.1.0))
+* Build from source : $ gradle createFatJar
```
change this to be:

./gradlew createFatJar

julie-sullivan commented on README.md:
````diff
@@ -62,4 +62,35 @@ Validator validator = new FastaValidator(isr, SequenceType.DNA);
+
+Setp 2: run as an executable jar file
+```shell
+java -jar biovalidator-0.1.0.jar --help
````
please update this!

Change to be one of these:

// run this in your build/libs directory
> java -jar biovalidator-0.1.2.jar --help

OR

> java -jar build/libs/biovalidator-0.1.2.jar --help

julie-sullivan replied:
biovalidator-fat-0.1.2.jar will be the name of the jar

julie-sullivan commented:
```
  -t, --type=<validatorType>
                            ValidatorType, fasta is used as default if not provided
```

Please change this. We do not want to default to fasta.

julie-sullivan commented:
```
julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar --file=ref_GRCh38.p12_top_level.gff3
Missing required option '--file=<filename>'
```

My syntax looks correct according to that message. Can you verify? Do I need quotes? The `-f` flag works just fine.

julie-sullivan commented:
```
julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar -f /tmp/data/variant_summary.txt
Validating  file...
Missing or Invalid Validator type! It must be one of ([FASTA, FASTA_DNA, FASTA_RNA, FASTA_PROTEIN, GFF, GFF3, CSV]), case-insensitive.
```
You can guess that a txt file will be CSV or TAB

julie-sullivan commented on src/main/java/org/intermine/biovalidator/CommandLineRunner.java:
```diff
@@ -57,27 +73,42 @@ public static void main(String[] args) {
+            }
+
+            //check --file arg is provided or not, as it is a required argument
+            if (!ArrayUtils.contains(args, "-f")) {
```
should check for `--file=<filename>` or `-f` I think?

julie-sullivan commented:
```
    /**
     * Return list of validation messages.
     *
     * @return Value of list of validation messages.
     */
    List<Message> getErrorMessages();

    /**
     * Returns a single error message as string describing the cause of validation failure
     * if validator failed or else it returns an empty string
     * @return string containing error message or empty string
     */
    String getErrorMessage();

```

What do i expect ` getErrorMessage()` to return if i have multiple errors?

All of the error messages concatenated? Or only the first error message? 

Can you update your docs to clarify? :)

deepakkumar96 commented:
Currently, It returns only the first error message.
Do you want this to return all errors concatenated?

On Mon, Aug 12, 2019 at 9:11 PM Julie Sullivan <notifications@github.com>
wrote:

>     /**
>      * Return list of validation messages.
>      *
>      * @return Value of list of validation messages.
>      */
>     List<Message> getErrorMessages();
>
>     /**
>      * Returns a single error message as string describing the cause of validation failure
>      * if validator failed or else it returns an empty string
>      * @return string containing error message or empty string
>      */
>     String getErrorMessage();
>
>
> What do i expect getErrorMessage() to return if i have multiple errors?
>
> All of the error messages concatenated? Or only the first error message?
>
> Can you update your docs to clarify? :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/biovalidator/pull/26?email_source=notifications&email_token=AED7EHGDGRI5QUZMMFS7YU3QEGADTA5CNFSM4IJGLJGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4C6BWA#issuecomment-520478936>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AED7EHCJ4V3IMPEHEFN6HQ3QEGADTANCNFSM4IJGLJGA>
> .
>

julie-sullivan commented:
no, whatever you want to do. :) but it wasn't clear from the docs what to expect.

deepakkumar96 commented:
So, I'm updating docs to clarify the behaviour.

On Mon, Aug 12, 2019 at 9:16 PM Julie Sullivan <notifications@github.com>
wrote:

> no, whatever you want to do. :) but it wasn't clear from the docs what to
> expect.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/biovalidator/pull/26?email_source=notifications&email_token=AED7EHFRIQW5BURTMIFHLWDQEGAWVA5CNFSM4IJGLJGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4C6SVA#issuecomment-520481108>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AED7EHANK66YMHQ4FSHE2HDQEGAWVANCNFSM4IJGLJGA>
> .
>

deepakkumar96 committed:
Fix review issues:
- change docs in README.MD
- change docs of ValidationResult::getErrorMessage
- Fix '--file' Missing error in CommandLineRunner
- Add "txt" as a validator-type when guessing validator type

julie-sullivan commented:
when validating tab file, says it's validating CSV file

```
julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar -f /tmp/data/variant_summary.tsv
Validating csv file...
```
You could just write "validating text file" :)

julie-sullivan commented:
I got 50 of the same error messages. I suppose you selected 50 as a cutoff on purpose? Just verifying!

And does that mean that the column does have values in it *some* times? If a column is always empty, does that generate a warning?

I think you have made all the right choices here! :) But in your report, for future work, I'd like you to do a good analysis on how we can extend your work here. What should we do next?

```
julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar -f /tmp/data/variant_summary.tsv -w
Validating csv file...
Result : Valid File!!
Warnings:
	1. column 26 at row 1 is blank
	2. column 26 at row 2 is blank
	3. column 26 at row 3 is blank
	4. column 26 at row 4 is blank
	5. column 26 at row 5 is blank
	6. column 26 at row 6 is blank
	7. column 26 at row 7 is blank
	8. column 26 at row 8 is blank
	9. column 26 at row 9 is blank
	10. column 26 at row 10 is blank
	11. column 26 at row 11 is blank
	12. column 26 at row 12 is blank
	13. column 26 at row 13 is blank
	14. column 26 at row 14 is blank
	15. column 26 at row 15 is blank
	16. column 26 at row 16 is blank
	17. column 26 at row 17 is blank
	18. column 26 at row 18 is blank
	19. column 26 at row 19 is blank
	20. column 26 at row 20 is blank
	21. column 26 at row 21 is blank
	22. column 26 at row 22 is blank
	23. column 26 at row 23 is blank
	24. column 26 at row 24 is blank
	25. column 26 at row 25 is blank
	26. column 26 at row 26 is blank
	27. column 26 at row 27 is blank
	28. column 26 at row 28 is blank
	29. column 26 at row 29 is blank
	30. column 26 at row 30 is blank
	31. column 26 at row 31 is blank
	32. column 26 at row 32 is blank
	33. column 26 at row 33 is blank
	34. column 26 at row 34 is blank
	35. column 26 at row 35 is blank
	36. column 26 at row 36 is blank
	37. column 26 at row 37 is blank
	38. column 26 at row 38 is blank
	39. column 26 at row 39 is blank
	40. column 26 at row 40 is blank
	41. column 26 at row 41 is blank
	42. column 26 at row 42 is blank
	43. column 26 at row 43 is blank
	44. column 26 at row 44 is blank
	45. column 26 at row 45 is blank
	46. column 26 at row 46 is blank
	47. column 26 at row 47 is blank
	48. column 26 at row 48 is blank
	49. column 26 at row 49 is blank
	50. column 26 at row 50 is blank
```

deepakkumar96 commented:
1. "when validating tab file, says it's validating CSV file"
 => The display message for csv, tsv, txt files is same,  'Validating csv
file...'
2. "I suppose you selected 50 as a cutoff on purpose?"                =>
Yes the max number warnings that can be generated is 50
3. " If a column is always empty, does that generate a warning?" => Yes
always report warning for empty column values.

On Tue, Aug 13, 2019 at 8:09 AM Julie Sullivan <notifications@github.com>
wrote:

> I got 50 of the same error messages. I suppose you selected 50 as a cutoff
> on purpose? Just verifying!
>
> And does that mean that the column does have values in it *some* times?
> If a column is always empty, does that generate a warning?
>
> I think you have made all the right choices here! :) But in your report,
> for future work, I'd like you to do a good analysis on how we can extend
> your work here. What should we do next?
>
> julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar -f /tmp/data/variant_summary.tsv -w
> Validating csv file...
> Result : Valid File!!
> Warnings:
> 	1. column 26 at row 1 is blank
> 	2. column 26 at row 2 is blank
> 	3. column 26 at row 3 is blank
> 	4. column 26 at row 4 is blank
> 	5. column 26 at row 5 is blank
> 	6. column 26 at row 6 is blank
> 	7. column 26 at row 7 is blank
> 	8. column 26 at row 8 is blank
> 	9. column 26 at row 9 is blank
> 	10. column 26 at row 10 is blank
> 	11. column 26 at row 11 is blank
> 	12. column 26 at row 12 is blank
> 	13. column 26 at row 13 is blank
> 	14. column 26 at row 14 is blank
> 	15. column 26 at row 15 is blank
> 	16. column 26 at row 16 is blank
> 	17. column 26 at row 17 is blank
> 	18. column 26 at row 18 is blank
> 	19. column 26 at row 19 is blank
> 	20. column 26 at row 20 is blank
> 	21. column 26 at row 21 is blank
> 	22. column 26 at row 22 is blank
> 	23. column 26 at row 23 is blank
> 	24. column 26 at row 24 is blank
> 	25. column 26 at row 25 is blank
> 	26. column 26 at row 26 is blank
> 	27. column 26 at row 27 is blank
> 	28. column 26 at row 28 is blank
> 	29. column 26 at row 29 is blank
> 	30. column 26 at row 30 is blank
> 	31. column 26 at row 31 is blank
> 	32. column 26 at row 32 is blank
> 	33. column 26 at row 33 is blank
> 	34. column 26 at row 34 is blank
> 	35. column 26 at row 35 is blank
> 	36. column 26 at row 36 is blank
> 	37. column 26 at row 37 is blank
> 	38. column 26 at row 38 is blank
> 	39. column 26 at row 39 is blank
> 	40. column 26 at row 40 is blank
> 	41. column 26 at row 41 is blank
> 	42. column 26 at row 42 is blank
> 	43. column 26 at row 43 is blank
> 	44. column 26 at row 44 is blank
> 	45. column 26 at row 45 is blank
> 	46. column 26 at row 46 is blank
> 	47. column 26 at row 47 is blank
> 	48. column 26 at row 48 is blank
> 	49. column 26 at row 49 is blank
> 	50. column 26 at row 50 is blank
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/biovalidator/pull/26?email_source=notifications&email_token=AED7EHGS72UPNORQ6UYCDRDQEJT45A5CNFSM4IJGLJGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4E4PZA#issuecomment-520734692>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AED7EHGHPS4HT4D75CWMUXTQEJT45ANCNFSM4IJGLJGA>
> .
>

julie-sullivan commented:
Does your code for tsv skip headers?

I can't totally tell, but I see the `hasHeader() iterating through the columns, trying to guess if there is a header? Is that right?

The alternative would be to always skip the first line. Or only skip lines with hashes. 

Something to discuss! :)

julie-sullivan commented:
1. "when validating tab file, says it's validating CSV file"
 => The display message for csv, tsv, txt files is same,  'Validating csv
file...'

Yes, I see that. It's incorrect, please fix.

julie-sullivan commented:
3. " If a column is always empty, does that generate a warning?" => Yes
always report warning for empty column values.

Maybe a future improvement (not for now! :)) is that we could just have one warning: "Column 6 is completely empty"

julie-sullivan commented:
2. "I suppose you selected 50 as a cutoff on purpose?"                =>
Yes the max number warnings that can be generated is 50

Maybe add this to the docs? Future work -- maybe allow this to be a param.

julie-sullivan commented on README.md:
````diff
@@ -62,4 +62,35 @@ Validator validator = new FastaValidator(isr, SequenceType.DNA);
+```
+
+### Using biovalidator as a command line utility
+Step 1: Get biovalidator Fat-Jar(either build from source or dirctly download from [bintray](https://bintray.com/intermineorg/biovalidator/biovalidator#files/org%2Fintermine%2Fbiovalidator%2F0.1.0))
````
1. you have bintray listed twice. remove one link
2. you don't mention the generated jar will be in `build/libs`. please add.

julie-sullivan commented on docs/Home.md:
```diff
@@ -0,0 +1,28 @@
+<dependency>
+  <groupId>org.intermine</groupId>
+  <artifactId>biovalidator</artifactId>
+  <version>0.1.0</version>
```
0.1.2

Safe to change, i think this will be our last release version?

julie-sullivan commented on docs/Home.md:
````diff
@@ -0,0 +1,28 @@
+
+Gradle dependency
+```groovy
+compile 'org.intermine:biovalidator:0.1.0'
````
0.1.2

julie-sullivan commented on src/main/java/org/intermine/biovalidator/CommandLineRunner.java:
```diff
@@ -137,30 +205,38 @@ private static void displayMessages(String messageType, List<Message> messages)
-                        + " possible values(fasta, fasta-dna, fasta-rna, fasta-protein)",
+                description = "ValidatorType, "
+                        + "possible values:\n fasta,\n fasta-dna,\n fasta-rna,\n fasta-protein,\n"
+                        + " gff3,\n csv",
```
Should tab / tsv be on this list?

deepakkumar96 replied:
should I add tsv? I thought CSV and TSV are the same types of file!

julie-sullivan replied:
yes, you should add TSV and TAB.

TSV is a tab separated file. TAB is a tab separated file. CSV is a comma separated file. They are different types of files.

The code may treat them similarly, but this is a text message for the end users. In this case, users will not consider tab and comma separated files the same types of files.

Does that make sense?

deepakkumar96 replied:
yes:)

julie-sullivan commented on src/main/java/org/intermine/biovalidator/api/DefaultValidationResult.java:
```diff
@@ -69,6 +84,11 @@ public boolean isValid() {
 
+    @Override
+    public boolean isNotValid() {
+        return !isValid;
```
should this return false?

deepakkumar96 replied:
isNotValid() is just a helper method so that I don't have to always negate isValid() method call while checking result is valid or Invalid

julie-sullivan commented on src/main/java/org/intermine/biovalidator/parser/CsvParser.java:
```diff
@@ -0,0 +1,107 @@
+
+    private com.univocity.parsers.csv.CsvParser createParser() {
+        CsvParserSettings settings = new CsvParserSettings();
+        if (hasHeader) {
```
how accurate is this? what's the tradeoff between just skipping the first line always?

(not proposing that!! Just wondering! :))

deepakkumar96 replied:
CsvHeaderDetector is an exact Java port of python std lib CSV module: here is the python docs : https://docs.python.org/3.1/library/csv.html#csv.Sniffer.has_header

julie-sullivan replied:
that's interesting.

So how accurate is it? What's it doing?

deepakkumar96 replied:
I won't say it's super smart, what it does is simple:
1. It looks for a column which is of a single type(say integer) and if all rows have the same type of value except the first row, then the first row is considered as a header.

2. If the first check doesn't work, then it also counts the length of the value of each column, if all values of a column have the same length except the first row, then again its a header.

So, I ported this python lib's algorithm to Java just because I thought this should work fine in most cases and the second reason is its part of python standard lib. But it can easily fail if all values in CSV file are strings with different length.

deepakkumar96 replied:
here's python documentation:
```pyton
 def has_header(self, sample):
        # Creates a dictionary of types of data in each column. If any
        # column is of a single type (say, integers), *except* for the first
        # row, then the first row is presumed to be labels. If the type
        # can't be determined, it is assumed to be a string in which case
        # the length of the string is the determining factor: if all of the
        # rows except for the first are the same length, it's a header.
        # Finally, a 'vote' is taken at the end for each column, adding or
        # subtracting from the likelihood of the first row being a header.
       ...
```

julie-sullivan commented:
testing the pattern matching.

1. 10 rows
2. first row is header with hash
3. first column is all integers

I changed the first row first column to be a string. I didn't get a warning. Is that expected?

julie-sullivan commented:
boolean pattern matching worked!

```
julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar -f /tmp/data/test.tsv -w -s
Validating csv file...
Result : Valid File!!
Warning:
	1. data is not consistent in column 2 and row 5, most rows have boolean but 1 rows has non-boolean values(s)
```

deepakkumar96 commented:
> when validating tab file, says it's validating CSV file
> 
> ```
> julie ~/git/biovalidator/build/libs $ java -jar biovalidator-fat-0.1.2.jar -f /tmp/data/variant_summary.tsv
> Validating csv file...
> ```
> 
> You could just write "validating text file" :)

this issue resolved!

deepakkumar96 committed:
Solve raised review issues:
- document CsvValidator code
- fix wrong error message in csv validator
- fix wrong BioValidator version in README.MD
- fix wrong validator type for txt files
